doc/api/util.md

```diff
@@ -1631,6 +1631,71 @@ const module = new WebAssembly.Module(wasmBuffer);
 util.types.isWebAssemblyCompiledModule(module);  // Returns true
 ```
 
+### util.createPromiseHook(hook)
+<!-- YAML
+added: REPLACEME
+-->
+
+* `hook` {Function}
+  * `type` {string}
+    * `resolve` Called when a promise resolve function is called. The promise
+      is not fulfilled or rejected at this point.
+    * `rejectWithNoHandler` Called when a promise rejects without any handler.
+    * `handlerAddedAfterReject` Called when a promise handler is added to a
+      promise that rejected without a handler.
+    * `rejectAfterResolved` Called when a promise reject function is called
+      after resolution.
+    * `resolveAfterResolved` Called when a promise resolve function is called
+      after resolution.
+  * `promise` {Promise}
+  * `value` {\*}
+
+Returns: {Object}
+  * `enable` {Function} Enables the hook
+  * `disable` {Function} Disables the hook
+
+Registers the promise hook `hook` to be called for certain promise debugging
+events.
+
+```js
+util.createPromiseHook((type, promise, value) => {
+  console.log('Promise did', type, 'with', value);
+}).enable();
```

Comment:
I would like to be able to receive only some of the events. Specifically `'resolve'` will call the hook all the time, creating a significant amount of overhead. At the moment adding this hook causes a 20% drop in throughput in some cases. I'd like to be able to notify `'rejectAfterResolved'` and `'resolveAfterResolved'` situations.

Reply:
I would like to be able to receive only some of the events. the event handler still has to handle and react to all the different event types, even if it doesn't fire for your handler. adding a filter wouldn't change perf.

Reply:
>  adding a filter wouldn't change perf.

I disagree. I think that if the filtering is done in C++ we are going to avoid the C++/JS boundary crossing for every resolved Promise, which is significant.

Reply:
If we're going with the filtering approach, then I'd like to see the API be more consistent with that used in async_hooks... specifically something like:

```js
const hook = util.createPromiseHook({
  resolve(promise, value) { /* .. */ },
  unhandledRejection(promise, value) { /* .. */ }
});
```

The idea is to keep diagnostics related APIs as similar and consistent as possible.

Reply:
this api was designed to be as stupidly close to the bare events as possible. i really don't want to abstract it like this, the filter already annoys me enough.

Reply:
Adding the filter solves the performance issue in having to listen to `'resolve'`, listening to `'rejectAfterResolved'` and `'resolveAfterResolved'` does not create any overhead at all.